The job **`aarch64-gnu-llvm-20-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/20666389074/job/59339483824) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/59339483824) [(plain)](https://github.com/rust-lang/rust/commit/8bf0ba19b4f6e5812c56e2a1ea16c582501f20cc/checks/59339483824/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
diff of stderr:

2   --> $DIR/long-conflict-markers.rs:3:1
3    |
4 LL | <<<<<<<<< HEAD
-    | ^^^^^^^^^ between this marker and `=======` is the code that we're merging into
+    | ^^^^^^^^^ between this marker and `=======` is the code that you are merging into
6 ...
7 LL | =========
8    | --------- between this marker and `>>>>>>>` is the incoming code

12    |
13    = note: conflict markers indicate that a merge was started but could not be completed due to merge conflicts
14            to resolve a conflict, keep only the code you want and then delete the lines containing conflict markers
-    = help: if you're having merge conflicts after pulling new code:
-            the top section is the code you already had and the bottom section is the remote code
-            if you're in the middle of a rebase:
-            the top section is the code being rebased onto and the bottom section is the code coming from the current commit being rebased
-    = note: for an explanation on these markers from the `git` documentation:
-            visit <https://git-scm.com/book/en/v2/Git-Tools-Advanced-Merging#_checking_out_conflicts>
+    = help: if you are in a merge, the top section is the code you already had checked out and the bottom section is the new code
+            if you are in a rebase, the top section is the code being rebased onto and the bottom section is the code you had checked out which is being rebased
+    = note: for an explanation on these markers from the `git` documentation, visit <https://git-scm.com/book/en/v2/Git-Tools-Advanced-Merging#_checking_out_conflicts>
21 
22 error: aborting due to 1 previous error
23 


The actual stderr differed from the expected stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args parser/diff-markers/long-conflict-markers.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/parser/diff-markers/long-conflict-markers.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/parser/diff-markers/long-conflict-markers" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
error: encountered diff marker
##[error]  --> /checkout/tests/ui/parser/diff-markers/long-conflict-markers.rs:3:1
   |
LL | <<<<<<<<< HEAD //~ ERROR encountered diff marker
   | ^^^^^^^^^ between this marker and `=======` is the code that you are merging into
...
LL | =========
   | --------- between this marker and `>>>>>>>` is the incoming code
LL |         y: i8,
LL | >>>>>>>>> branch
   | ^^^^^^^^^ this marker concludes the conflict region
   |
   = note: conflict markers indicate that a merge was started but could not be completed due to merge conflicts
           to resolve a conflict, keep only the code you want and then delete the lines containing conflict markers
   = help: if you are in a merge, the top section is the code you already had checked out and the bottom section is the new code
           if you are in a rebase, the top section is the code being rebased onto and the bottom section is the code you had checked out which is being rebased
   = note: for an explanation on these markers from the `git` documentation, visit <https://git-scm.com/book/en/v2/Git-Tools-Advanced-Merging#_checking_out_conflicts>

error: aborting due to 1 previous error
------------------------------------------

---- [ui] tests/ui/parser/diff-markers/long-conflict-markers.rs stdout end ----
```

</details>

--- a/compiler/rustc_parse/src/parser/diagnostics.rs
+++ b/compiler/rustc_parse/src/parser/diagnostics.rs
@@ -3037,27 +3037,53 @@ impl<'a> Parser<'a> {
         long_kind: &TokenKind,
         short_kind: &TokenKind,
     ) -> bool {
-        (0..3).all(|i| self.look_ahead(i, |tok| tok == long_kind))
-            && self.look_ahead(3, |tok| tok == short_kind)
+        if long_kind == short_kind {
+            // For conflict marker chars like `%` and `\`.
+            (0..7).all(|i| self.look_ahead(i, |tok| tok == long_kind))
+        } else {
+            // For conflict marker chars like `<` and `|`.
+            (0..3).all(|i| self.look_ahead(i, |tok| tok == long_kind))
+                && self.look_ahead(3, |tok| tok == short_kind || tok == long_kind)
+        }
     }
 
-    fn conflict_marker(&mut self, long_kind: &TokenKind, short_kind: &TokenKind) -> Option<Span> {
+    fn conflict_marker(
+        &mut self,
+        long_kind: &TokenKind,
+        short_kind: &TokenKind,
+        expected: Option<usize>,
+    ) -> Option<(Span, usize)> {
         if self.is_vcs_conflict_marker(long_kind, short_kind) {
             let lo = self.token.span;
-            for _ in 0..4 {
-                self.bump();
+            if self.psess.source_map().span_to_margin(lo) != Some(0) {
+                return None;
             }
-            return Some(lo.to(self.prev_token.span));
+            let mut len = 0;
+            while self.token.kind == *long_kind || self.token.kind == *short_kind {
+                if self.token.kind.break_two_token_op(1).is_some() {
+                    len += 2;
+                } else {
+                    len += 1;
+                }
+                self.bump();
+                if expected == Some(len) {
+                    break;
+                }
+            }
+            if expected.is_some() && expected != Some(len) {
+                return None;
+            }
+            return Some((lo.to(self.prev_token.span), len));
         }
         None
     }
 
     pub(super) fn recover_vcs_conflict_marker(&mut self) {
         // <<<<<<<
-        let Some(start) = self.conflict_marker(&TokenKind::Shl, &TokenKind::Lt) else {
+        let Some((start, len)) = self.conflict_marker(&TokenKind::Shl, &TokenKind::Lt, None) else {
             return;
         };
-        let mut spans = Vec::with_capacity(3);
+        let mut spans = Vec::with_capacity(2);
         spans.push(start);
         // |||||||
         let mut middlediff3 = None;
@@ -3069,13 +3095,19 @@ impl<'a> Parser<'a> {
             if self.token == TokenKind::Eof {
                 break;
             }
-            if let Some(span) = self.conflict_marker(&TokenKind::OrOr, &TokenKind::Or) {
+            if let Some((span, _)) =
+                self.conflict_marker(&TokenKind::OrOr, &TokenKind::Or, Some(len))
+            {
                 middlediff3 = Some(span);
             }
-            if let Some(span) = self.conflict_marker(&TokenKind::EqEq, &TokenKind::Eq) {
+            if let Some((span, _)) =
+                self.conflict_marker(&TokenKind::EqEq, &TokenKind::Eq, Some(len))
+            {
                 middle = Some(span);
             }
-            if let Some(span) = self.conflict_marker(&TokenKind::Shr, &TokenKind::Gt) {
+            if let Some((span, _)) =
+                self.conflict_marker(&TokenKind::Shr, &TokenKind::Gt, Some(len))
+            {
                 spans.push(span);
                 end = Some(span);
                 break;

--- a/tests/ui/parser/diff-markers/long-conflict-markers.rs
+++ b/tests/ui/parser/diff-markers/long-conflict-markers.rs
@@ -1,0 +1,11 @@
+enum E {
+    Foo {
+<<<<<<<<< HEAD //~ ERROR encountered diff marker
+        x: u8,
+|||||||
+        z: (),
+=========
+        y: i8,
+>>>>>>>>> branch
+    }
+}

--- a/tests/ui/parser/diff-markers/long-conflict-markers.stderr
+++ b/tests/ui/parser/diff-markers/long-conflict-markers.stderr
@@ -1,0 +1,20 @@
+error: encountered diff marker
+  --> $DIR/long-conflict-markers.rs:3:1
+   |
+LL | <<<<<<<<< HEAD
+   | ^^^^^^^^^ between this marker and `=======` is the code that you are merging into
+...
+LL | =========
+   | --------- between this marker and `>>>>>>>` is the incoming code
+LL |         y: i8,
+LL | >>>>>>>>> branch
+   | ^^^^^^^^^ this marker concludes the conflict region
+   |
+   = note: conflict markers indicate that a merge was started but could not be completed due to merge conflicts
+           to resolve a conflict, keep only the code you want and then delete the lines containing conflict markers
+   = help: if you are in a merge, the top section is the code you already had checked out and the bottom section is the new code
+           if you are in a rebase, the top section is the code being rebased onto and the bottom section is the code you had checked out which is being rebased
+   = note: for an explanation on these markers from the `git` documentation, visit <https://git-scm.com/book/en/v2/Git-Tools-Advanced-Merging#_checking_out_conflicts>
+
+error: aborting due to 1 previous error
+

